src/device-state/current-state.ts

```diff
@@ -138,128 +135,109 @@ const getStateDiff = (): DeviceStatus => {
 	};
 
 	return _.omitBy(diff, _.isEmpty);
-};
-
-const throttledReport = _.throttle(
-	// We define the throttled function this way to avoid UncaughtPromise exceptions
-	// for exceptions thrown from the report function
-	(opts: StateReport, resolve: () => void, reject: (e: Error) => void) =>
-		report(opts).then(resolve).catch(reject),
-	constants.maxReportFrequency,
-);
+}
 
-/**
- * Perform exponential backoff on the function, increasing the attempts
- * counter on each call
- *
- * If attempts is 0 then, it will delay for min(minDelay, maxDelay)
- */
-const backoff = (
-	retry: (attempts: number) => void,
-	attempts = 0,
-	maxDelay: number,
-	minDelay = MINIMUM_BACKOFF_DELAY,
-) => {
-	const delay = Math.min(2 ** attempts * minDelay, maxDelay);
-	log.info(`Retrying request in ${delay / 1000} seconds`);
-	setTimeout(() => retry(attempts + 1), delay);
-};
+async function reportCurrentState(conf: CurrentStateReportConf) {
+	// Ensure no other report starts
+	reportPending = true;
+	// Wrap the report with fetching of state so report always has the latest state diff
+	const wrappedReport = async () => {
+		// Get state to report
+		const stateToReport = await generateStateForReport();
+		// Get diff from last reported state
+		const stateDiff = newStateDiff(stateToReport);
+		// Report diff
+		if (await report({ stateDiff, conf })) {
+			// Update lastReportedState
+			_.assign(lastReportedState.local, stateDiff.local);
+			_.assign(lastReportedState.dependent, stateDiff.dependent);
+			// Log that we successfully reported the current state
+			log.info('Reported current state to the cloud');
+		}
+	};
+	// Create a report that will backoff on errors
+	const reportWithBackoff = withBackoff(wrappedReport, {
+		maxDelay: conf.appUpdatePollInterval,
+		minDelay: 15000, // If no retryAfter passed, start at 15 seconds
+		retryCallback: handleRetry,
+	});
+	// Run in try block to avoid throwing any exceptions
+	try {
+		await reportWithBackoff();
+	} catch (e) {
+		log.debug(e);
+	}
+	stateReportErrors = 0;
+	reportPending = false;
+}
 
-function reportCurrentState(attempts = 0) {
-	(async () => {
-		const {
-			hardwareMetrics,
-			appUpdatePollInterval: maxDelay,
-			...conf
-		} = await config.getMany([
-			'deviceApiKey',
-			'apiTimeout',
-			'apiEndpoint',
-			'uuid',
-			'localMode',
-			'appUpdatePollInterval',
-			'hardwareMetrics',
-		]);
+function handleRetry(retryInfo: BackoffCallback) {
+	if (retryInfo.error instanceof StatusError) {
+		log.error(
+			`Device state report failure! Status code: ${retryInfo.error.statusCode} - message:`,
+			retryInfo.error.message,
```

Comment:
```suggestion
			retryInfo.error?.message ?? retryInfo.error,
```
In case an API response doesn't contain a message